.circleci/config.yml

```diff
@@ -207,19 +207,99 @@ commands:
       - attach_workspace:
           at: /home/circleci/
 
+  # For the restore-deps-cache commands below, only restore the cache if there's an exact match. This means whatever
+  # is in the cache will be exactly what's used and the cache won't keep growing uncontrollably going forward.
+
   restore-be-deps-cache:
     steps:
       - restore_cache:
           keys:
-            - be-deps-v4-{{ checksum "project.clj" }}
-            - be-deps-v4-
+            - be-deps-v5-{{ checksum "project.clj" }}
 
   restore-fe-deps-cache:
     steps:
       - restore_cache:
           keys:
-            - fe-deps-v4-{{ checksum "yarn.lock" }}
-            - fe-deps-v4-
+            - fe-deps-v5-{{ checksum "yarn.lock" }}
+
+  # Only run steps if changes have happened since the last time this was ran successfully (determined by the cache key)
+  # Cache key should be unique for job and relevant source files -- use a checksum.
+  #
+  # Important! If this step is skipped because no changes have happened, the entire JOB will halt with a success
+  # status -- no steps that happen AFTER run-on-change will be ran. Keep this in mind!
+  run-on-change:
+    parameters:
+      cache-key:
+        type: string
+      steps:
+        type: steps
+    steps:
+      - restore_cache:
+          keys:
+            - run-on-change-<< parameters.cache-key >>
+      - run:
+          name: Skip rest of job if .SUCCESS exists for cache key << parameters.cache-key >>
+          command: |
+            if [ -f .SUCCESS ]; then
+                echo "Job previously succeeded for cache key << parameters.cache-key >>"
+                echo "Link to last successful run: $(cat .SUCCESS)"
+                circleci-agent step halt
```

Comment:
should the `.SUCCESS` file be removed at this point to prevent further steps from thinking that they also have a cache key? I see at the bottom of this step it is removed but not sure how far reaching the `circleci-agent step halt` command is here.

Reply:
`circleci-agent step halt` immediately stops the job and returns a successful result. So there's no situation where `.SUCCESS` can end up making it anywhere else, since nothing in that job will be run after `circleci-agent step halt`

Reply:
oh right. I'm thinking the tests are run in succession in the same job but that's not the case at all. thanks